arrow-libs/plugins/src/main/kotlin/arrow/gradle/MppPublishingPlugin.kt

```diff
@@ -0,0 +1,159 @@
+  override fun apply(target: Project) {}
+}
+
+fun Project.setupPublishing(
+  pomDevId: String,
+  pomDevName: String,
+  projectUrl: String,
+  projectDesc: String,
+  releaseRepo: java.net.URI,
+  snapshotRepo: java.net.URI,
+  sonatypeUsername: String? = System.getenv("SONATYPE_USER"),
+  sonatypePassword: String? = System.getenv("SONATYPE_PWD"),
+  pass: String? = System.getenv("SIGNINGPASSWORD"),
+  key: String? = System.getenv("SIGNINGKEY"),
+  licenseName: String = "The Apache Software License, Version 2.0",
+  licenseUrl: String = "https://www.apache.org/licenses/LICENSE-2.0.txt",
+  licenseDist: String = "repo",
+  scmConnection: String = "${projectUrl.replaceFirst("https://", "scm:git:git://")}.git",
+  scmDevConnection: String = "${projectUrl.replaceFirst("https://", "scm:git:ssh://")}.git",
+) {
+  afterEvaluate {
+    extensions.getByType(PublishingExtension::class.java).apply {
+      val mavenPublications = publications.withType<MavenPublication>()
+      mavenPublications.all {
+        artifact(project.javadocJar())
+        setupPom(
+          url = projectUrl,
+          description = projectDesc,
+          pomDevId = pomDevId,
+          pomDevName = pomDevName,
+          licenseName = licenseName,
+          licenseUrl = licenseUrl,
+          licenseDist = licenseDist,
+          scmConnection = scmConnection,
+          scmDevConnection = scmDevConnection
+        )
+      }
+      repositories {
+        maven(
+          if (version.toString().endsWith("SNAPSHOT")) snapshotRepo else releaseRepo,
+          sonatypeUsername,
+          sonatypePassword
+        )
+      }
+      Nullable.zip(key, pass) { key, pass -> signPublications(key, pass) }
+    }
+  }
+}
+
+object Nullable {
+  fun <A : Any, B : Any, C : Any> zip(a: A?, b: B?, f: (A, B) -> C?): C? =
+    a?.let { aa ->
+      b?.let { bb -> f(aa, bb) }
+    }
+}
+
+fun Project.signPublications(
+  key: String,
+  pass: String
+) {
+  val publications = project
+    .extensions
+    .getByType(PublishingExtension::class.java)
+    .publications
+
+  project
+    .extensions
+    .getByType(SigningExtension::class.java)
+    .apply {
+      useInMemoryPgpKeys(key, pass)
+    }.sign(publications)
+}
+
+/* We either try to find the existing javadocJar, or we register an empty javadocJar task */
+fun Project.javadocJar(): TaskProvider<Jar> {
+  val taskName = "javadocJar"
+  return try {
+    tasks.named(name = taskName)
+  } catch (e: UnknownTaskException) {
+    tasks.register(name = taskName) {
+      archiveClassifier by "javadoc"
+    }
+  }
+}
+
+fun MavenPublication.setupPom(
+  url: String,
+  description: String,
+  pomDevId: String,
+  pomDevName: String,
+  licenseName: String,
+  licenseUrl: String,
+  licenseDist: String,
+  scmConnection: String,
+  scmDevConnection: String,
+) {
+  pom {
+    if (!name.isPresent) {
+      name by artifactId
+    }
+    this@pom.description.set(description)
+    this@pom.url.set(url)
```

Comment:
I think we may not need the tags on all the `this` in this function

Reply:
We do because it conflicts with the params of the function... Better/different paramnames for the functions? :/